shared-libs/transitions/src/transitions/create_user_for_contacts.js

```diff
@@ -127,6 +128,26 @@ const getUsersToReplace = (originalContact) => {
     .map(([username, { replacement_contact_id }]) => ({ username, replacementContactId: replacement_contact_id }));
 };
 
+const addUser = async (contact) => {
+  if (
+    !contact.role &&
+    (!Array.isArray(contact.roles) || !contact.roles.length)
+  ) {
+    throw new Error(`Contact [${contact._id}] must have a "role" or "roles" property.`);
+  }
+
+  const roles = Array.isArray(contact.roles) && contact.roles.length > 0 ? contact.roles : [contact.role];
+  await createNewUser({ roles }, contact);
+  delete contact.user_for_contact.create;
+};
+
+const isCreatingUser = (contact, info) => contact.user_for_contact.create === 'true'
+  && !transitionUtils.hasRun(info, NAME)
+  && info.initial_replication_date === info.latest_replication_date;
```

Comment:
Well, shoot!  For some reason I thought that `latest_replication_date` was set in Sentinel...

Unfortunately, I don't think we can just check for the existence of `transitions` since that value seems to be [set automatically](https://github.com/medic/cht-core/blob/a43f4fb3c32fb4286d4a939c01fa3457516824ae/shared-libs/infodoc/src/infodoc.js#L97) when you get the info doc...

Reply:
This is insane.  As far as I can tell there is no way to reliably check if a doc has been seen by Sentinel before!  I think we are either going to have to build that functionality out or we just need to give up on the requirement that you can only add users for new contacts (and just let you add users for existing contacts too).  I am going to take this question back to the issue....

Reply:
That code does not run in API, fortunately. 
These are the only two functions that are called by API: https://github.com/medic/cht-core/blob/a43f4fb3c32fb4286d4a939c01fa3457516824ae/shared-libs/infodoc/src/infodoc.js#L290-L291

I think it's possible to add something that will indicate that this is the first time sentinel has ever processed the document. If you can't leverage something that already exists, just add something new. Luckily this doesn't have to be backwards compatible with anything, since you don't care about docs created prior to this change to be evaluated as new.

Reply:
> That code does not run in API, fortunately.

I am not sure why API matters though, since it [gets run by Sentinel](https://github.com/medic/cht-core/blob/d3ee100dc0fb27347c48b8126ed78c788ae234cf/shared-libs/transitions/src/transitions/index.js#L53) before executing the transitions.  Either way, we `transitions` is not going to be a viable check and Rube Goldberg is proud....

I guess I can "solve" my problem by making another random edit to the info doc in Sentinel to set some magic flag.  :facepalm:

Reply:
> random edit to the info doc in Sentinel to set some magic flag.

I think that if you add the "flag" here, it won't even ever get saved: https://github.com/medic/cht-core/blob/a43f4fb3c32fb4286d4a939c01fa3457516824ae/shared-libs/infodoc/src/infodoc.js#L137

because we only ever update specific properties of the infodocs.

Reply:
Either way this a useful addition, it even seems silly you _can't_ tell whether this is the first time Sentinel has processed a certain document. 
You could even consider adding a specific date field to the infodoc with `processed_by_sentinel`.

Reply:
Okay, so here is the approach that I am trying:
- Update `infodoc` to not auto-populate `info.transitions` when calling `infodoc.get`
- That means that when the transitions lib _starts_ processing a doc change it can know if it has processed the doc before based on the presence of `info.transitions`
    - Based on that check, the transitions lib can set a `change.initialProcessing` flag which will allow the individual transitions to know if this is the first time the doc is being processed by Sentinel
- Once the transitions lib finishes processing a change it will save the `infodoc.transitions` value to the DB so that subsequent runs can properly set the `initialProcessing` flag to false

I like this approach because the behavior of `infodoc.get` is less surprising. We don't end up adding any extra flags to the info doc in the DB (beyond the existing `transitions`). And the `initialProcessing` flag logic can be localized to the transitions util.

Reply:
I think this is good. I don't think we actually need the transitions property before writing the first transition result, and all the rest of the code should already account for a missing property, since it initially does not exist in the infodoc.